x/shield/keeper/proposal.go

```diff
@@ -4,16 +4,153 @@ import (
+
+	"github.com/certikfoundation/shentu/x/shield/types"
 )
 
-func (k Keeper) ClaimLock(ctx sdk.Context, proposalID, poolID uint64, purchaser sdk.AccAddress, purchaseID uint64, loss sdk.Coins, lockPeriod time.Duration) error {
+// SecureCollaterals is called after a claim is submitted to secure
+// the given amount of collaterals for the duration and adjust shield
+// module states accordingly.
+func (k Keeper) SecureCollaterals(ctx sdk.Context, poolID uint64, purchaser sdk.AccAddress, purchaseID uint64, loss sdk.Coins, duration time.Duration) error {
+	lossAmt := loss.AmountOf(k.sk.BondDenom(ctx))
+
+	// Verify shield.
+	pool, found := k.GetPool(ctx, poolID)
+	if !found {
+		return types.ErrNoPoolFound
+	}
+	if lossAmt.GT(pool.Shield) {
+		return types.ErrNotEnoughShield
+	}
+
+	// Verify collateral availability.
+	totalCollateral := k.GetTotalCollateral(ctx)
+	totalClaimed := k.GetTotalClaimed(ctx)
+	totalClaimed = totalClaimed.Add(lossAmt)
+	if totalClaimed.GT(totalCollateral) {
+		panic("total claimed surpassed total collateral")
+	}
+
+	// Secure the updated loss ratio from each provider.
+	providers := k.GetAllProviders(ctx)
+	lossRatio := totalClaimed.ToDec().Quo(totalCollateral.ToDec())
+	for i := range providers {
+		lockAmt := providers[i].Collateral.ToDec().Mul(lossRatio).TruncateInt()
+		if lockAmt.LT(providers[i].Collateral) {
+			lockAmt = lockAmt.Add(sdk.OneInt())
+		}
+		k.SecureFromProvider(ctx, providers[i], lockAmt, duration)
+	}
+
+	// Update purchase states.
+	purchaseList, found := k.GetPurchaseList(ctx, poolID, purchaser)
+	if !found {
+		return types.ErrPurchaseNotFound
+	}
+	var index int
+	for i, entry := range purchaseList.Entries {
+		if entry.PurchaseID == purchaseID {
+			index = i
+			break
+		}
+	}
+	purchase := &purchaseList.Entries[index]
+	if lossAmt.GT(purchase.Shield) {
+		return types.ErrNotEnoughShield
+	}
+	purchase.Shield = purchase.Shield.Sub(lossAmt)
+	votingEndTime := ctx.BlockTime().Add(duration)
+	if purchase.DeletionTime.Before(votingEndTime) {
+		purchase.DeletionTime = votingEndTime
+	}
+	k.SetPurchaseList(ctx, purchaseList)
+
+	// Update pool and global pool states.
+	pool.Shield = pool.Shield.Sub(lossAmt)
+	k.SetPool(ctx, pool)
+
+	totalShield := k.GetTotalShield(ctx)
+	totalShield = totalShield.Sub(lossAmt)
+	k.SetTotalShield(ctx, totalShield)
+	k.SetTotalClaimed(ctx, totalClaimed)
+
 	return nil
 }
 
-func (k Keeper) ClaimUnlock(ctx sdk.Context, proposalID, poolID uint64, loss sdk.Coins) error {
-	return nil
+// SecureFromProvider secures the specified amount of collaterals from
+// the provider for the duration. If necessary, it extends withdrawing
+// collaterals and, if exist, their linked unbondings as well.
+func (k Keeper) SecureFromProvider(ctx sdk.Context, provider types.Provider, amount sdk.Int, duration time.Duration) {
+	// If there are enough bonded delegations backing
+	// locked collaterals, we are done.
+	if provider.DelegationBonded.GTE(amount) {
```

Comment:
Considering delegation bonded is not good enough.

Example:
I'm the only one in the shield pool. I staked 100CTK and deposited 100CTK. Then I withdraw all my collateral immediately after someone purchased a shield without undelegating my delegations. When the purchaser file a claim on Day 21, my withdraw will not be delayed because I have enough delegations. However, the purchaser will not be able to find anyone to pay his loss because my withdraw has been finished.

Reply:
Right Thanks. I should have been more careful there 🤔 I think I can just remove these incorrect checks?

Reply:
Sounds good.